src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java

```diff
@@ -0,0 +1,347 @@
+		}
+
+ 		@Nullable RemotePlayer remotePlayer = mRemotePlayersByUuid.get(playerUuid);
+		if (remotePlayer == null) {
+			return null;
+		}
+		return remotePlayer.mName;
+	}
+
+	protected @Nullable UUID getPlayerUuid(@Nullable String playerName) {
+		if (playerName == null) {
+			return null;
+		}
+
+		@Nullable RemotePlayer remotePlayer = mRemotePlayersByName.get(playerName);
+		if (remotePlayer == null) {
+			return null;
+		}
+		return remotePlayer.mUuid;
+	}
+
+	protected void refreshLocalPlayers() {
+		for (Player player : Bukkit.getOnlinePlayers()) {
+			refreshLocalPlayer(player);
+		}
+	}
+
+	protected void refreshLocalPlayer(Player player) {
+		mPlugin.getLogger().fine("Refreshing local player");
+		RemotePlayer remotePlayer = new RemotePlayer(player, true);
+
+		unregisterPlayer(remotePlayer.mUuid);
+		mPlugin.getLogger().fine("Registering local player");
+		mRemotePlayersByUuid.put(remotePlayer.mUuid, remotePlayer);
+		mRemotePlayersByName.put(remotePlayer.mName, remotePlayer);
+		remotePlayer.broadcast();
+	}
+
+	private void unregisterPlayer(UUID playerUuid) {
+		@Nullable RemotePlayer lastPlayerState = mRemotePlayersByUuid.get(playerUuid);
+		if (lastPlayerState != null) {
+			mPlugin.getLogger().fine("Unregistering local player");
+			String lastLoc = lastPlayerState.mShard;
+			@Nullable Map<String, RemotePlayer> lastShardRemotePlayers = mRemotePlayerShardMapped.get(lastLoc);
+			if (lastShardRemotePlayers != null) {
+				lastShardRemotePlayers.remove(lastPlayerState.mName);
+			}
+			mRemotePlayersByUuid.remove(playerUuid);
+			mRemotePlayersByName.remove(lastPlayerState.mName);
+
+			RemotePlayerUnloadedEvent event = new RemotePlayerUnloadedEvent(lastPlayerState);
+			Bukkit.getServer().getPluginManager().callEvent(event);
+		}
+	}
+
+	private void remotePlayerChange(JsonObject data) {
+		RemotePlayer remotePlayer;
+		try {
+			remotePlayer = new RemotePlayer(data);
+		} catch (Exception ex) {
+			mPlugin.getLogger().warning("Received invalid RemotePlayer");
+			return;
+		}
+
+		unregisterPlayer(remotePlayer.mUuid);
+		if (remotePlayer.mIsOnline) {
+			@Nullable Map<String, RemotePlayer> shardRemotePlayers = mRemotePlayerShardMapped.get(remotePlayer.mShard);
+			if (shardRemotePlayers != null) {
+				shardRemotePlayers.put(remotePlayer.mName, remotePlayer);
+			}
+			mPlugin.getLogger().fine("Registering local player from remote data");
+			mRemotePlayersByUuid.put(remotePlayer.mUuid, remotePlayer);
+			mRemotePlayersByName.put(remotePlayer.mName, remotePlayer);
+
+			RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
+			Bukkit.getServer().getPluginManager().callEvent(remotePLE);
+		} else if (!Objects.equals(getShardName(), remotePlayer.mName)) {
+			mPlugin.getLogger().fine("Detected race condition, triggering refresh on " + remotePlayer.mName);
+			@Nullable Player localPlayer = Bukkit.getPlayer(remotePlayer.mUuid);
+			if (localPlayer != null) {
+				refreshLocalPlayer(localPlayer);
+			}
+		}
```

Comment:
This logic doesn't make sense to me - why are we comparing the shard name and the player name?

Reply:
doesn't make sense to me, but if you don't have it, when someone changes shard, the remote update arrives before the local quit event (resulting in the entire entry being deleted everywhere)

Reply:
I just noticed that I was checking the player name